opt/scopeserver/dataserver/utils/search.py

```diff
@@ -0,0 +1,267 @@
+            matches = [x for x in self.search_space.keys() if self.casefold_term == x[0]]
+        else:
+            matches = [x for x in self.search_space.keys() if self.casefold_term in x[0]]
+
+        match_results = []
+
+        for match in matches:
+            if isinstance(self.search_space[match], list):
+                for element in self.search_space[match]:
+                    match_results.append((match, (match[1], element, match[2])))
+            else:
+                match_results.append((match, (match[1], self.search_space[match], match[2])))
+
+        perfect_match: List[SearchMatch] = []
+        good_match: List[SearchMatch] = []
+        bad_match: List[SearchMatch] = []
+
+        for match_result in match_results:
+            # This contains the full match, it's category and their relationship
+            match, result = match_result
+
+            if match[1] == self.search_term or match[0] == self.casefold_term:
+                perfect_match.append(result)
+                continue
+            if (
+                match[1].startswith(self.search_term)
+                or match[1].endswith(self.search_term)
+                or match[0].startswith(self.casefold_term)
+                or match[0].endswith(self.casefold_term)
+            ):
+                good_match.append(result)
+                continue
+            if self.search_term in match[0]:
+                bad_match.append(result)
+                continue
+
+        def sort_matches(match: SearchMatch) -> str:
+            return match[0]
+
+        final_matches = (
+            sorted(perfect_match, key=sort_matches)
+            + sorted(good_match, key=sort_matches)
+            + sorted(bad_match, key=sort_matches)
+        )
+
+        return final_matches
+
+    def aggregate_matches(self, matches: List[SearchMatch]) -> Dict[Tuple[str, str], List[str]]:
+        """
+        Collapse matches based on returned element.
+
+        match[0]: Match from search space
+        match[1]: Category of match
+        match[2]: Relationship
+        """
+
+        aggregated_matches: Dict[Tuple[str, str], List[str]] = OrderedDict()
+
+        if self.cross_species == "":
+            for match in matches:
+                key = (match[1], match[2])
+                if key not in aggregated_matches.keys():
+                    aggregated_matches[key] = [match[0]]
+                else:
+                    aggregated_matches[key].append(match[0])
+
+        elif self.cross_species == "hsap":
+            for match in matches:
+                key = (match[1], match[2])
+                for droso_gene in self.dfh.hsap_to_dmel_mappings[match[0]]:
+                    if key not in aggregated_matches.keys():
+                        aggregated_matches[key] = [droso_gene[0]]
+                    else:
+                        aggregated_matches[key].append(droso_gene[0])
+                    self.cross_species_identity[match[0]] = droso_gene[1]
```

Comment:
`match[0]` isn't being updated in this loop. This should be updated outside of this function anyway, it's not "aggregating matches"

Reply:
If I split the identity extraction out of here, we'll have to re-loop through the matches, would you prefer I rename the function to `process_and_aggregate_matches` or something like that?

Reply:
In principle I would prefer you loop through the matches again. Is this likely to be expensive? If it's expensive this name change is fine.